fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java

```diff
@@ -0,0 +1,359 @@
+                String indexIdsString = parameter.getValue().as(com.ibm.fhir.model.type.String.class).getValue();
+                if (indexIdsString != null) {
+                    String[] indexIdsArray = indexIdsString.split(",");
+                    for (String indexId : indexIdsArray) {
+                        boolean queued = false;
+                        while (!queued && this.running) {
+                            try {
+                                queued = blockingQueue.offer(indexId, OFFER_TIMEOUT_IN_SEC, TimeUnit.SECONDS);
+                            } catch (InterruptedException e) {
+                                // NOP
+                            }
+
+                            if (queued) {
+                                lastIndexId = indexId;
+                            } else {
+                                logger.warning("Unable to add indexId '" + indexId + "' to queue");
+                                if (!this.active) {
+                                    logger.warning("Worker threads are not active yet, so try adding again later");
+                                    return true;
+                                }
+                            }
+                        }
+                    }
+                    return indexIdsArray.length > 0;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Thread to repeatedly call the $reindex operation until done or error.
+     */
+    private void callReindexOperationInLoop() {
+        while (this.running && this.active) {
+            String indexIds = getIndexIdsToReindex();
+            if (!indexIds.isEmpty()) {
+                boolean ok = false;
+                try {
+                    ok = callReindexOperation(indexIds);
+                } catch (DataAccessException x) {
+                    // allow active be set to false.  This will notify monitorLoop something is wrong.
+                    // Probably all threads will encounter the same exception and monitorLoop will
+                    // try to refill the pool if all threads exit.
+                    logger.severe("DataAccessException caught when contacting FHIR server. FHIR client thread will exit." + x.toString() );
+                } catch (IllegalStateException x) {
+                    // Fail for this exception too. fhir-bucket fhir client suggests this exception results from config error.
+                    // So probably this will be caught first time monitorLoop calls callOnce and not here.
+                    logger.severe("IllegalStateException caught. FHIR client thread will exit." + x.toString() );
+                }
+                if (!ok) {
+                    // stop everything on the first failure
+                    this.active = false;
+                }
+            } else {
+                // queue is empty, wait a partial second and try again
+                safeSleep(100);
```

Comment:
This could limit throughput if threads are frequently sleeping waiting for new work. Better to use locks and conditions.

Reply:
Replaced this with a call to blockingQueue.poll(...), so it doesn't use this arbitrary sleep to wait to check if the queue is still empty.